session/kx.go

```diff
@@ -0,0 +1,439 @@
+// Copyright (c) 2016,2017 Company 0, LLC.
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package session
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/sha512"
+	"errors"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/companyzero/sntrup4591761"
+	"github.com/davecgh/go-xdr/xdr2"
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+var (
+	ErrDecrypt   = errors.New("decrypt failure")
+	ErrOverflow  = errors.New("message too large")
+	ErrInvalidKx = errors.New("invalid kx")
+	ErrMarshal   = errors.New("could not marshal")
+	ErrUnmarshal = errors.New("could not unmarshal")
+)
+
+// KX allows two peers to derive a pair of shared keys. One peer must trigger
+// Initiate (the client) while the other (the server) should call Init once
+// followed by Respond for each connection.
+type KX struct {
+	Conn           net.Conn
+	MaxMessageSize uint
+	OurPrivateKey  *[sntrup4591761.PrivateKeySize]byte
+	OurPublicKey   *[sntrup4591761.PublicKeySize]byte
+	TheirPublicKey *[sntrup4591761.PublicKeySize]byte
+	writeKey       *[32]byte
+	readKey        *[32]byte
+	writeSeq       [24]byte
+	readSeq        [24]byte
+}
+
+// A pair of ephemeral keys is kept to ensure key erasure (forward secrecy)
+// should long-term keys be compromised.
+var (
+	ephemeralPublic  [sntrup4591761.PublicKeySize]byte
+	ephemeralPrivate [sntrup4591761.PrivateKeySize]byte
+	ephemeralMutex   sync.Mutex
+)
+
+// regenerateEphemeral rotates the server/client's ephemeral key. It is invoked
+// concurrently to the operation of the server/client, therefore we need to
+// acquire a mutex to ensure noninterference.
+func regenerateEphemeral() error {
+	pk, sk, err := sntrup4591761.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	ephemeralMutex.Lock()
+	copy(ephemeralPublic[:], pk[:])
+	copy(ephemeralPrivate[:], sk[:])
+	ephemeralMutex.Unlock()
+	return nil
+}
+
+// zeroEphemeral erases the contents of ephemeralP{ublic,rivate}.
+func zeroEphemeral() {
+	for i := range ephemeralPublic {
+		ephemeralPublic[i] ^= ephemeralPublic[i]
+	}
+	for i := range ephemeralPrivate {
+		ephemeralPrivate[i] ^= ephemeralPrivate[i]
+	}
+}
+
+// Init prepares the server to start responding to kx initiation requests.
+// It calls regenerateEphemeral once, and then once every minute. If we fail
+// to rotate our ephemeral key, we bring the server down.
+func Init() {
```

Comment:
I really would like this to be inside a context vs being global.

Reply:
Hi Marco,

This function is indeed a bit odd, particularly since it is only used
by the server. We could perhaps abstract it to a 'session.Responder'
type, which would give it a context.

However, I believe this is out of the scope of the PR, and something
we should tackle in the refactoring that the zkc codebase so badly
needs.